neural_network_lyapunov/relu_to_optimization.py

```diff
@@ -561,6 +562,50 @@ def _compute_layer_bound(self,
                             x_lo.detach().numpy(),
                             x_up.detach().numpy(),
                             create_prog_callback, binary_var_type)
+                if method == mip_utils.PropagateBoundsMethod.IA_MIP:
+                    # We also compute the bounds by IA. If the IA bound is
+                    # close to the MIP bound, then we use the IA bound.
+                    # Otherwise, we relax the MIP bound a little bit to make
+                    # sure this relaxed bound will not be active.
+                    z_pre_relu_lo_ia, z_pre_relu_up_ia = \
+                        mip_utils.compute_range_by_IA(
+                            self.model[2 * layer_count].weight, bias,
+                            linear_layer_input_lo, linear_layer_input_up)
+                    for j in range(self.model[2 * layer_count].out_features):
+                        neuron_index = self.relu_unit_index[layer_count][j]
+                        if torch.abs(z_pre_relu_lo[neuron_index] -
```

Comment:
I'm definitely a little bit worried that this depends on the tolerance we use on the solver for the MIP bound. If we ever play with that to improve efficiency it might bite us. I think it's fine, but something to keep in mind.

Reply:
It is better to use a tolerance larger than the solver accuracy. So I use 1E-4 here. It is OK to use the IA bounds, which carries the right gradient. It would be bad if a bound could be active, but we use the MIP bound which drops the gradient. Hence a larger tolerance can avoid such problem.